### PR TITLE
chore: bump to node 21.1.0

### PR DESCRIPTION
**Pull request by TimBeyer**
**What this PR does / why we need it**:
Does what it says.

See https://github.com/TypeStrong/ts-node/issues/1997 for the reason for the `ts-node` change even with `esm: true` being set.

**Which issue(s) this PR fixes**:

Closes https://github.com/garden-io/garden/issues/4158

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by shumailxyz** [comment]
This should close issue https://github.com/garden-io/garden/issues/4158, right?

**Comment by TimBeyer** [comment]
There's an issue with `ts-node` which if fixed using the suggested `node --experimental-loader=ts-node/esm` causes terrible performance for everything run through `ts-node`.
The performance is so bad, that I wouldn't even consider merging it even if it's just impacting CI script times.

We'd need to either wait for a fix in the project, or maybe see if we can somehow skip running things with `ts-node`.

**Comment by TimBeyer** [comment]
The new fix seems to be better so I guess we're good to go with this.

File: scripts/register-hook.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import url from 'node:url'
```
Comment (stefreak) [inline scripts/register-hook.js:1]: It might be good to add a comment here explaining why this file exists.

**Review by stefreak — APPROVED** [review]
👍

**Pull request merged**